### ISSUE TITLE
Ocusync 3

### ISSUE THREAD

**Issue by lobart** [proto17/dji_droneid#25]
Hi, i followed your wiki for mavic 3.
And when i did bruteforce for root, i found 147 and 600 roots, not 385.
![147](https://user-images.githubusercontent.com/26677576/200313187-f9c87547-f2c8-45b6-89b0-778a37e50b4f.png)
![600](https://user-images.githubusercontent.com/26677576/200313205-9931ad39-571a-4b42-ba11-e44c0d263952.png)

And graph with 147 root shows that we have symbols with width 256 samples.
And i don't know, why there are not 385 root.
I can send you samples of mavic 3 and mavic 3 mini pro

**Comment by proto17** [comment]
Hmm, your results also show that each ZC sequence is only sent once which is not what I found (each sequence was repeated once and both root indices were sent back to back).  The collect I used to find the different sequences was provided to me by someone else.  So I don't know exactly what drone was used.  Perhaps there is a firmware difference?

So long as you're okay with the fact that your location might be present I'd be very happy to have a look.  One thing that would be very interesting is to see a power spectrum (waterfall) of any drones you have access to.

Thanks!

**Comment by lobart** [comment]
@proto17 I got this results from mavic 3 and mavic 3 mini pro. Both bursts shown, that part of algorithm for ocusync 2 before quak demodulation does not get good result for  quantizing. I used several cycle_prefixes, but results not good too. I show you tomorrow

**Comment by lobart** [comment]
@proto17 well, as i said
This is burst of mavic3:
![image](https://user-images.githubusercontent.com/26677576/200510962-5478c813-ce92-4756-a398-3a1b87ae4074.png)

This is burst of mavic 3 mini pro:
![image](https://user-images.githubusercontent.com/26677576/200511184-9380d99c-04b3-4d67-a622-572ff0921756.png)

And all of this bursts get two root indices 147 and 600.

But prior qpsk quantization OFDM-symbols look like this:
![80_72_1_foo](https://user-images.githubusercontent.com/26677576/200511800-5c16b99b-89ea-4ee4-97cc-cc8228e485c8.png)

When symbols of Ocusync 2 look like this:
![34_72_foo](https://user-images.githubusercontent.com/26677576/200511914-906afb3d-4e0f-4075-aec5-f6020bd7934d.png)

I tried to use brute force for cyclic prefix from 1 to 100, and filter taps from 1 to 50, but a could not get result like from ocusync2.

**Comment by proto17** [comment]
@lobart thanks for the images!  It looks like both adhere to the frame structure that is currently used in the code base: 2 data symbols, 1 ZC, 1 data symbol, 1 ZC, 3 data symbols.  The ZC root indices look the same too.  You can fingerprint the root index fairly easily by looking at the angle and number of stripes in the waterfall.  Not sure why the first won't lock or why the second one doesn't look right in the 6th and 9th symbols.

One thing that's known to really goof up demod is a large center freq offset.  I added logic to the `main` [branch](../commit/95bc95e236f42d14c4a43e62d019723610f23024) to help with detecting the integer frequency offset (IFO).  Was that being used with the files that you processed?  If so, what was the value of `integer_offset`?

**Comment by lobart** [comment]
There are orange plot behind blue by 3 samples.
![image](https://user-images.githubusercontent.com/26677576/201360361-19da78c7-8156-4d33-8967-e0f709792edb.png)

**Comment by proto17** [comment]
@lobart something goofy is happening and i don't have a good feel for exactly what.  if you're okay with sending me the file, what is your preferred way to accomplish that?

**Comment by proto17** [comment]
@lobart one thing that might be worth trying is changing [line 168 of process_file.m](https://github.com/proto17/dji_droneid/blob/main/matlab/updated_scripts/process_file.m#L168) to something like `5` or `10`.  I didn't realize that I committed the code with no interpolation which is a problem.  This mistake means that there is a good chance for a large fractional STO which isn't great.  Also, make sure you're using `main` and not any of the other branches as `main` has the most up-to-date MATLAB.

**Comment by proto17** [comment]
@lobart after tinkering with the `process_file.m` script to force it to create a cyclic prefix plot like yours i think i know what's wrong:  Double check that your sample rate is correct.  the only way that I could recreate your issue was to resample the file to something like `15.38e6` or `30.74e6`.  i can't think of another reason that the cyclic prefix plots would lag like that.

**Comment by lobart** [comment]
> @lobart something goofy is happening and i don't have a good feel for exactly what. if you're okay with sending me the file, what is your preferred way to accomplish that?
@proto17 
@boltraa in telegram

**Comment by lobart** [comment]
> @lobart after tinkering with the `process_file.m` script to force it to create a cyclic prefix plot like yours i think i know what's wrong: Double check that your sample rate is correct. the only way that I could recreate your issue was to resample the file to something like `15.38e6` or `30.74e6`. i can't think of another reason that the cyclic prefix plots would lag like that.

@proto17 I used fft_size 1021 and got result 
![image](https://user-images.githubusercontent.com/26677576/201619403-aa17b9a9-f7b5-4c4e-b547-916b193ba69c.png)

**Comment by lobart** [comment]
@proto17 I added 1 to long_cyclic_prefix
![image](https://user-images.githubusercontent.com/26677576/201631052-291925a4-c63d-4b33-89b4-086e67a7018d.png)
Can you describe numbers at 7 and 8 lines?

**Comment by proto17** [comment]
That's just a way to get the cyclic prefix lengths based on the sample rate.  What I probably should have done is:

```matlab
long_cp_len = round(sample_rate * 5.2e-6);
short_cp_len = round(sample_rate * 4.69e-6);
```

The `5.2e-6` is 5.2 microseconds, and `4.69e-6` is 4.69 microseconds.  Both values came from [this](https://www.sharetechnote.com/html/Handbook_LTE_PhyParameter_DL_FDD.html) page.

Did you happen to use a non-standard rate (ie not 15.36e6 or 30.72e6)?  What SDR did you use?  The existing code would likely not work for a non-standard rate.  You can try replacing what's in the repo with the code above.

**Comment by lobart** [comment]
@proto17 I use b210 usrp with 15.36M rate
But with mavic 2 it works.
See i processed burst by hand. And i choose fft_size 1021 and angles for phase for every OFDM-symbol.
And i got this:
![image](https://user-images.githubusercontent.com/26677576/201682962-17ab53e7-4a7e-4b67-958e-25ddb8f65fe5.png)

**Comment by lobart** [comment]
@proto17 I used your script with ocusync 2, and it doesnt work. But previous version of your script is correct.

**Comment by catkira** [comment]
if you have to lower the fft_size from 1024 to 102,1 it means that you have a STO of about 3 samples per 1024 samples (0.3%). The spectrum of the signal appears to small for you, which means that your sample rate is a bit to high. It should work with the original fft_size if you lower the sample rate about 0.3%. 
Do you use an original b210 or a china clone? Does it have the GPSDO option?

**Comment by lobart** [comment]
@catkira I use original board. 
https://www.ettus.com/all-products/ub210-kit/
So, do you mean that I should to use 15.14M sample rate instead 15.36M ?

**Comment by catkira** [comment]
yes, and also try + 0.3 %, because the fact that you have to decrease fft_size to make it work also hints that your sample rate could be too slow. Because by shortening the fft_size you also reduce the length of each symbol in time.

**Comment by catkira** [comment]
I would write a script that plots the distances of the CPs for the entire burst. From the cp distance you can tell what your sample rate offset is.

**Comment by lobart** [comment]
@catkira I don’t understand, what do you mean under distances of the CP? 
I did brute force CPs from 1 to 100, but lag was the same (3 samples)

**Comment by catkira** [comment]
Normally there should be a CP (short or long) every 1024 samples after the last CP. You can verify that thats the case using cross correlation. 

You can also plot the phase of the channel estimation. If the phase has a slope thats a sign of STO. I would use the channel estimation of the first ZC symbol for that.

Its a bit difficult to explain it all with words. Can you provide your burst file? I can write a small analysis in python then when I have time (probably this weekend).

**Comment by lobart** [comment]
@catkira you can send me your mail to lobachav@gmail.com , and I will send you file
